// standardized-audio-context/build/es2018/factories/minimal-base-audio-context-constructor.js
import { EventTarget } from '../event-target';
import { CONTEXT_STORE } from '../globals';
export const createMinimalBaseAudioContextConstructor = (audioDestinationNodeConstructor, createAudioListener) => {
    return class MinimalBaseAudioContext extends EventTarget {
        constructor(nativeContext, numberOfChannels) {
            super();
            CONTEXT_STORE.set(this, nativeContext);
            // Bug #93: Edge will set the sampleRate of an AudioContext to zero when it is closed.
            const sampleRate = nativeContext.sampleRate;
            Object.defineProperty(nativeContext, 'sampleRate', {
                get: () => sampleRate
            });
            this._destination = new audioDestinationNodeConstructor(this, numberOfChannels);
            this._listener = createAudioListener(this, nativeContext);
            this._nativeContext = nativeContext;
        }
        get currentTime() {
            return this._nativeContext.currentTime;
        }
        get destination() {
            return this._destination;
        }
        get listener() {
            return this._listener;
        }
        get onstatechange() {
            return this._nativeContext.onstatechange;
        }
        set onstatechange(value) {
            this._nativeContext.onstatechange = value;
        }
        get sampleRate() {
            return this._nativeContext.sampleRate;
        }
        get state() {
            return this._nativeContext.state;
        }
    };
};
