// standardized-audio-context/build/es2018/factories/wave-shaper-node-renderer-factory.js
import { getNativeAudioNode } from '../helpers/get-native-audio-node';
import { isOwnedByContext } from '../helpers/is-owned-by-context';
import { renderInputsOfAudioNode } from '../helpers/render-inputs-of-audio-node';
export const createWaveShaperNodeRendererFactory = (createNativeWaveShaperNode) => {
    return () => {
        let nativeWaveShaperNode = null;
        return {
            render: async (proxy, nativeOfflineAudioContext) => {
                if (nativeWaveShaperNode !== null) {
                    return nativeWaveShaperNode;
                }
                nativeWaveShaperNode = getNativeAudioNode(proxy);
                /*
                 * If the initially used nativeWaveShaperNode was not constructed on the same OfflineAudioContext it needs to be created
                 * again.
                 */
                if (!isOwnedByContext(nativeWaveShaperNode, nativeOfflineAudioContext)) {
                    const options = {
                        channelCount: nativeWaveShaperNode.channelCount,
                        channelCountMode: nativeWaveShaperNode.channelCountMode,
                        channelInterpretation: nativeWaveShaperNode.channelInterpretation,
                        curve: nativeWaveShaperNode.curve,
                        oversample: nativeWaveShaperNode.oversample
                    };
                    nativeWaveShaperNode = createNativeWaveShaperNode(nativeOfflineAudioContext, options);
                }
                await renderInputsOfAudioNode(proxy, nativeOfflineAudioContext, nativeWaveShaperNode);
                return nativeWaveShaperNode;
            }
        };
    };
};
